standalone/regex.c

```diff
@@ -1,162 +1,237 @@
+
+preg->re_pcre = pcre2_compile(pcre2_pattern, PCRE2_ZERO_TERMINATED,
+  options, &error_number, &error_offset, NULL);
+preg->re_erroffset = error_offset;
+
+if (preg->re_pcre == NULL) return AP_REG_INVARG;
+
+pcre2_pattern_info((const pcre2_code *)preg->re_pcre, PCRE2_INFO_CAPTURECOUNT, &nsub);
+preg->re_nsub = nsub;
+
+#else
+if ((cflags & AP_REG_ICASE) != 0) options |= PCRE_CASELESS;
+if ((cflags & AP_REG_NEWLINE) != 0) options |= PCRE_MULTILINE;
+
+preg->re_pcre = pcre_compile(pattern, options, &errorptr, &erroffset, NULL);
+preg->re_erroffset = erroffset;
+
+if (preg->re_pcre == NULL) return AP_REG_INVARG;
+
+pcre_fullinfo((const pcre *)preg->re_pcre, NULL, PCRE_INFO_CAPTURECOUNT, &nsub);
+preg->re_nsub = nsub;
+#endif  // end of WITH_PCRE
+return 0;
+}
+
+#ifndef POSIX_MALLOC_THRESHOLD
+#define POSIX_MALLOC_THRESHOLD (10)
+#endif
+
+AP_DECLARE(int) ap_regexec(const ap_regex_t *preg, const char *string,
+                           apr_size_t nmatch, ap_regmatch_t pmatch[],
+                           int eflags)
+{
+int rc;
+int options = 0;
+int *ovector = NULL;
+int small_ovector[POSIX_MALLOC_THRESHOLD * 3];
+int allocated_ovector = 0;
+
+#ifdef WITH_PCRE2
+if ((eflags & AP_REG_NOTBOL) != 0) options |= PCRE2_NOTBOL;
+if ((eflags & AP_REG_NOTEOL) != 0) options |= PCRE2_NOTEOL;
+#else
+if ((eflags & AP_REG_NOTBOL) != 0) options |= PCRE_NOTBOL;
+if ((eflags & AP_REG_NOTEOL) != 0) options |= PCRE_NOTEOL;
+#endif
+
+((ap_regex_t *)preg)->re_erroffset = (apr_size_t)(-1);  /* Only has meaning after compile */
+
+if (nmatch > 0)
+  {
+  if (nmatch <= POSIX_MALLOC_THRESHOLD)
+    {
+    ovector = &(small_ovector[0]);
+    }
+  else
+    {
+    ovector = (int *)malloc(sizeof(int) * nmatch * 3);
+    if (ovector == NULL) return AP_REG_ESPACE;
+    allocated_ovector = 1;
+    }
+  }
+
+#ifdef WITH_PCRE2
+{
+  PCRE2_SPTR pcre2_s;
+  int pcre2_ret;
+  pcre2_match_data *match_data;
+  PCRE2_SIZE *pcre2_ovector = NULL;
+
+  pcre2_s = (PCRE2_SPTR)string;
+  match_data = pcre2_match_data_create_from_pattern(preg->re_pcre, NULL);
+  pcre2_match_context *match_context = pcre2_match_context_create(NULL);
+
+  pcre2_ret = pcre2_match((const pcre2_code *)preg->re_pcre, pcre2_s, (int)strlen(string),
+      0, (uint32_t)options, match_data, match_context);
+
+  if (match_data != NULL) {
+      pcre2_ovector = pcre2_get_ovector_pointer(match_data);
+      if (pcre2_ovector != NULL) {
+          for (int i = 0; ((i < pcre2_ret) && ((i*2) <= nmatch * 3)); i++) {
+              if ((i*2) < nmatch * 3) {
+                  ovector[2*i] = pcre2_ovector[2*i];
+                  ovector[2*i+1] = pcre2_ovector[2*i+1];
+              }
+          }
+      }
+      pcre2_match_data_free(match_data);
+      pcre2_match_context_free(match_context);
+  }
+}
+#else
+rc = pcre_exec((const pcre *)preg->re_pcre, NULL, string, (int)strlen(string),
+  0, options, ovector, nmatch * 3);
+#endif
+
+if (rc == 0) rc = nmatch;    /* All captured slots were filled in */
```

Comment:
I don's see `rc` beeing set in the PCRE2 case. It looks to me like the code in the following couple of lines shouldn't run in case of PCRE2.

Reply:
You're right. I assumed `rc` is initialized with `0` when it's declared, but that's sure. See c2216b26 where I set it up (with the given comment).